### DISCUSSION TITLE
How to make a Timer Interrupt version of blink LED

### DISCUSSION THREAD

**Discussion by MrYorkiebar** [earlephilhower/arduino-pico#589] [Q&A] [answered]
I am trying to learn how to use a timer interrupt on the pico to make a basic interrupt version of blink. I'm currently very stuck because I can't find resources about timers and interrupts using C++ and this core on the Arduino IDE or PlatformIO. I have programmed interrupts on the mega 2560 using the port registers like in this [post](https://www.instructables.com/Arduino-Timer-Interrupts/) but it appears to be different on the pico. Considering this I have no idea how to get interrupts working and I keep hitting a dead end with googling. When googling I keep getting either micropython (not C++) or using the Pico C SDK which doesn't seem to work how I thought. I also can't find any timer interrupt stuff in the documentation for this core located [here](https://arduino-pico.readthedocs.io/en/latest/search.html?q=timer&check_keywords=yes&area=default).
I have tried this on the arduino IDE:
```
#include <stdio.h>
#include "pico/stdlib.h"

// the setup function runs once when you press reset or power the board

bool LED_state = false;

bool flash_LED_callback(struct repeating_timer *t){
  LED_state = !LED_state;
  digitalWrite(LED_BUILTIN, LED_state);
  return true;
}

void setup() {
  // initialize digital pin LED_BUILTIN as an output.
  pinMode(LED_BUILTIN, OUTPUT);
  // initialize a timer
  struct repeating_timer timer;
  add_repeating_timer_ms(-1000, flash_LED_callback, NULL, &timer)
}

// the loop function runs over and over again forever
void loop() {
  // digitalWrite(LED_BUILTIN, HIGH);   // turn the LED on (HIGH is the voltage level)
  // delay(1000);                       // wait for a second
  // digitalWrite(LED_BUILTIN, LOW);    // turn the LED off by making the voltage LOW
  // delay(1000);                       // wait for a second
}

bool banana = 1;
void setup1() {
  banana = 0;
}

// the loop function runs over and over again forever
void loop1() {
  banana = !banana;
}
```

which is based of of [this ](https://raspberrypi.github.io/pico-sdk-doxygen/group__hardware__timer.html)example code and the basic blink code. I thought that "A complete copy of the PICO SDK is included with the Arduino core, and all functions in the core are available inside the standard link libraries." meant that I could just do #include "pico/stdlib.h" but this just says not found. I've only really installed libraries using the library manager in the Arduino IDE or VSCode with platformIO so I don't really know how it works.

I'm probably misunderstanding something but I don't know what. Any help on how to use timer interrupts on the pico would be really helpful? 

My future goal is to use this interrupt to drive stepper motor step and direction pins for 6 to 8 motors with complex movement.

**Comment by khoih-prog** [comment, marked as answer]
Try my [**RPI_PICO_TimerInterrupt** library](https://github.com/khoih-prog/RPI_PICO_TimerInterrupt), and possible [**RP2040_PWM**](https://github.com/khoih-prog/RP2040_PWM) for future PWM-related motor control.

**Reply by MrYorkiebar** [reply]
Thank you, this looks good, I will look into this and post a follow up.

**Reply by MrYorkiebar** [reply]
I tried your library and it too had the error of #include "pico/stdlib.h" not being found but it also suggested that I was not using the correct core and it prompted me to check the selected board. It turns out I had chosen "Raspberry Pi Pico (Picoprobe)" so I switched to "Raspberry Pi Pico" and it worked. I have used you library in the following blink code:
```
#include <RPi_Pico_TimerInterrupt.h>

// Setup interval between switching the LED on or off
#define FLASH_INTERVAL_US_SLOW     1000000
#define FLASH_INTERVAL_US_FAST      100000

// Initialise Timers
RPI_PICO_Timer myTimer0(0);
RPI_PICO_Timer myTimer1(1);

// Global Variables
bool fastFlash = true;
bool LED_state = false;

// Timer interrupt service routines
bool flash_LED_callback_SLOW(struct repeating_timer *t){
  LED_state = !LED_state;
  digitalWrite(LED_BUILTIN, LED_state);
  return true;
}
bool flash_LED_callback_FAST(struct repeating_timer *t){
  LED_state = !LED_state;
  digitalWrite(LED_BUILTIN, LED_state);
  return true;
}

void setup() {
  // Initialise serial
  Serial.begin(9600);

  // initialize digital pin LED_BUILTIN as an output.
  pinMode(LED_BUILTIN, OUTPUT);
  
  // Interval in microsecs
  myTimer0.attachInterruptInterval(FLASH_INTERVAL_US_SLOW, flash_LED_callback_SLOW);
  myTimer1.attachInterruptInterval(FLASH_INTERVAL_US_FAST, flash_LED_callback_FAST);
  myTimer1.stopTimer();
}
void setup1() {
  fastFlash = false;  // Arbitrary use of setup1 just for fun
}

// Loop for alternating between the timers
void loop() {
  delay(5000);  //Wait 5 seconds before switching

  // Swap the timers
  if (fastFlash){
    // Stop fast flash and start slow flash
    myTimer1.stopTimer();
    myTimer0.restartTimer();
  }else{
    // Stop slow flash and start fast flash
    myTimer0.stopTimer();
    myTimer1.restartTimer();
  }
  fastFlash = !fastFlash;
}
void loop1() {
  // Print current state to serial
  if (fastFlash){
    Serial.println("Fast flash. Timer0 disabled Timer1 enabled");
  }else{
    Serial.println("Slow flash. Timer1 disabled Timer0 enabled");
  }
}
```
Your library helped me create a simple interrupt driven blink LED program and also the error reporting helped me notice the issue with the wrong board selected so I've marked it as the answer :smile: .
